File: 0257-binary-tree-paths/0257-binary-tree-paths.java
class Solution {
    private List<String> result;

    private void dfs(TreeNode node, String route) {
        if (node.left == null && node.right == null) {
            result.add(route);
            return;
        }
        if (node.left != null) {
            dfs(node.left, route + "->" + node.left.val);
        }
        if (node.right != null) {
            dfs(node.right, route + "->" + node.right.val);
        }
    }

    public List<String> binaryTreePaths(TreeNode root) {
        result = new ArrayList<>();
        dfs(root, Integer.toString(root.val));
        return result;
    }
}

File: 0257-binary-tree-paths/0257-binary-tree-paths.cpp
class Solution {
    vector<string> result;

    void dfs(TreeNode* node, string route) {
        route += to_string(node->val);
        if (node->left == NULL && node->right == NULL) {
            result.push_back(route);
            return;
        }
        route += "->";
        if (node->left != NULL) {
            dfs(node->left, route);
        }
        if (node->right != NULL) {
            dfs(node->right, route);
        }
    }

public:
    vector<string> binaryTreePaths(TreeNode* root) {
        dfs(root, "");
        return result;
    }
};